### PR TITLE
Refactored Tink Worker

### PR DESCRIPTION
**Pull request by micahhausler**
## Description

* Externalized worker package for testability  
* Created `ContainerManager` and `ContainerLogger` abstractions
* Added tests for new interfaces
* Cleaned up worker method arguments (logger is now passed via context)
* Left worker logic unchanged
* Refactored `Worker` struct and initialization

## Why is this needed

I was working on #573 and wanted to be able to create a 'virtual worker' with faked-out Docker client calls. This refactor will make it easy immediately test API changes and later refactor the worker logic.

## How Has This Been Tested?

* Tested locally with a workflow
* Added unit tests

## How are existing users impacted? What migration steps/scripts do we need?

No impact to existing users

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [x] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

File: cmd/tink-worker/worker/container_manager.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+func (m *containerManager) WaitForFailedContainer(ctx context.Context, id string, failedActionStatus chan pb.State) {
+	l := m.getLogger(ctx)
+	// send API call to wait for the container completion
+	wait, errC := m.cli.ContainerWait(ctx, id, container.WaitConditionNotRunning)
+
+	select {
+	case status := <-wait:
+		if status.StatusCode == 0 {
+			failedActionStatus <- pb.State_STATE_SUCCESS
+			return
```
Comment (micahhausler) [inline cmd/tink-worker/worker/container_manager.go:131]: This method is only ever called in a goroutine, but if it isn't, this code will block on a non-buffered channel so I added the return here

File: cmd/tink-worker/worker/container_manager_test.go [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+			registry:    "rebelba.se",
+			clientErr:   errors.New("You missed the shot"),
+			wantErr:     errors.New("DOCKER CREATE: You missed the shot"),
+		},
+	}
+	logger, err := log.Init("github.com/tinkerbell/tink")
+	if err != nil {
+		t.Errorf("Error initializing logger: %w", err)
+		return
+	}
```
Comment (mmlb) [inline cmd/tink-worker/worker/container_manager_test.go:134]: you're probably better of using https://pkg.go.dev/github.com/packethost/pkg@v0.0.0-20211110202003-387414657e83/log#Test instead, it has the benefit of only printing output when `test.v` is enabled or if the test fails.

File: cmd/tink-worker/worker/registry.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/docker/docker/api/types"
+	"github.com/pkg/errors"
+)
+
+// RegistryConnDetails are the connection details for accessing a Docker
+// registry.
+type RegistryConnDetails struct {
+	Registry,
+	User,
+	Pwd string
```
[the remark this answers is not in the document]
Comment (mmlb) [reply cmd/tink-worker/worker/registry.go:18]: can we also do `Username` and `Password` please, they read easier and (*bonus*) are the same length!

File: cmd/tink-worker/worker/worker.go [hunk 4]
```diff
@@ -376,14 +425,15 @@ func isLastAction(wfContext *pb.WorkflowContext, actions *pb.WorkflowActionList)
...
-	l := w.logger.With("workflowID", actionStatus.GetWorkflowId,
+	l := w.getLogger(ctx).With("workflowID", actionStatus.GetWorkflowId(),
 		"workerID", actionStatus.GetWorkerId(),
 		"actionName", actionStatus.GetActionName(),
 		"taskName", actionStatus.GetTaskName(),
 	)
 	var err error
 	for r := 1; r <= w.retries; r++ {
-		_, err = w.client.ReportActionStatus(ctx, actionStatus)
+		l.Info("Reporting Action Status")
```
Comment (mmlb) [inline cmd/tink-worker/worker/worker.go:435]: log messages should not be `Sentence case`. I thought we had a linter to catch this :thinking: (maybe needs a lint-install update). Would be nice not to introduce cases where we'd need to fix later on.

**Review by mmlb — COMMENTED** [review]
Looks really good, thanks for the work. I've only got minor issue with a few things, please see the comments.

File: cmd/tink-worker/worker/container_logger.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	"fmt"
+	"io"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/packethost/pkg/log"
+)
+
+// ContainerLogger emits container logs.
+type ContainerLogger interface {
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply cmd/tink-worker/worker/container_logger.go:15]: How about `LogCapturer`?

File: cmd/tink-worker/worker/registry.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	"encoding/base64"
+	"encoding/json"
+	"io"
+
+	"github.com/docker/docker/api/types"
+	"github.com/pkg/errors"
+)
+
+// RegistryConnDetails are the connection details for accessing a Docker
+// registry.
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply cmd/tink-worker/worker/registry.go:14]: will fix, copied over

File: cmd/tink-worker/worker/container_manager.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	wfDir := filepath.Join(defaultDataDir, wfID)
+	hostConfig := &container.HostConfig{
+		Privileged: privileged,
+		Binds:      []string{wfDir + ":/workflow"},
+	}
+
+	if pidConfig := action.GetPid(); pidConfig != "" {
+		l.With("pid", pidConfig).Info("creating container")
```
[the remark this answers is not in the document]
Comment (micahhausler) [reply cmd/tink-worker/worker/container_manager.go:82]: I'll just drop this line

File: cmd/tink-worker/worker/container_manager_test.go (hunk 2 above)
Comment (micahhausler) [reply cmd/tink-worker/worker/container_manager_test.go:134]: nice

File: cmd/tink-worker/worker/worker.go (hunk 4 above)
Comment (micahhausler) [reply cmd/tink-worker/worker/worker.go:435]: fixed

File: cmd/tink-worker/worker/container_logger.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mmlb) [reply cmd/tink-worker/worker/container_logger.go]: sounds less awkward yeah :D

File: cmd/tink-worker/worker/log_capturer_test.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+			content: "Line1\nline2\n",
+		},
+		{
+			name:    "empty buffer from error",
+			writer:  *bytes.NewBufferString(""),
+			wanterr: errors.New("Docker failure"),
+			content: "",
+		},
+	}
+	logger := log.Test(t, "github.com/tinkerbell/tink")
```
Comment (mmlb) [inline cmd/tink-worker/worker/log_capturer_test.go:57]: ditto

Comment (micahhausler) [reply cmd/tink-worker/worker/log_capturer_test.go]: fixed

**Comment by micahhausler** [comment]
@mmlb Fixed logger initialization

File: cmd/tink-worker/worker/log_capturer.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	"fmt"
+	"io"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/packethost/pkg/log"
+)
+
+// LogCapturer emits container logs.
+type LogCapturer interface {
```
Comment (chrisdoherty4) [inline cmd/tink-worker/worker/log_capturer.go:15]: What consumes this interface? Nothing in this package seems to so I wonder if we could just export the struct and leave the interface declaration to consumers?

Comment (micahhausler) [reply cmd/tink-worker/worker/log_capturer.go:15]: Sure, I can make the implementation public. I think I'll still leave the interface in this package unless there is a reason not to.

Comment (chrisdoherty4) [reply cmd/tink-worker/worker/log_capturer.go:15]: https://github.com/golang/go/wiki/CodeReviewComments#interfaces

Comment (micahhausler) [reply cmd/tink-worker/worker/log_capturer.go:15]: I'll move the `LogCapturer` and `ContainerManager` interface declarations into `worker.go`. They're still in the same package, but will be declared at the site of use.

**Comment by micahhausler** [comment]
@mmlb anything else on this?

File: cmd/tink-worker/cmd/root.go [hunk 10]
```diff
@@ -3,16 +3,18 @@ package cmd
 import (
 	"context"
 	"fmt"
+	"os"
 	"strings"
 	"time"
 
+	dockerCli "github.com/docker/docker/client"
```
Comment (chrisdoherty4) [inline cmd/tink-worker/cmd/root.go:10]: Relatively minor: alias's should be all lowercase same as package names.

Comment (micahhausler) [reply cmd/tink-worker/cmd/root.go]: Fixed

**Review by thebsdbox — DISMISSED** [review]
Fantastic, thanks for the additional testing !

File: cmd/tink-worker/worker/container_manager.go [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+	"github.com/packethost/pkg/log"
+	"github.com/pkg/errors"
+	pb "github.com/tinkerbell/tink/protos/workflow"
+)
+
+const (
+	ErrFailedToWait   = "failed to wait for completion of action"
+	ErrFailedToRunCmd = "failed to run on-timeout command"
+
+	InfoWaitFinished = "wait finished for failed or timeout container"
```
Comment (mmlb) [inline cmd/tink-worker/worker/container_manager.go:20]: I don't think these need to/should be public right? They're only used within this package.

**Review by mmlb — COMMENTED** [review]
lgtm, just the one comment about the public strings that seem unnecessary.

File: cmd/tink-worker/worker/container_manager.go (hunk 11 above)
Comment (micahhausler) [reply cmd/tink-worker/worker/container_manager.go:20]: I can make them private

**Review by mmlb — APPROVED** [review]

**Pull request merged**